### ISSUE TITLE
[Bug]: canvas pick events not working when Axes belongs to a subfigure

### ISSUE THREAD

**Issue by ndeadly** [matplotlib/matplotlib#25860]
### Bug summary

When an axis is added to a subfigure, rather than the root figure, canvas pick events are no longer received for that axis.

### Code for reproduction

```python
import matplotlib.pyplot as plt
import matplotlib.gridspec as mgridspec
import numpy as np

def on_pick(evt):
    print('pick event!')

fig = plt.figure()
gs = mgridspec.GridSpec(1, 1)

# Pick events working
#ax = fig.add_subplot(gs[0, 0])

# Pick events not working
subfig = fig.add_subfigure(gs[0, 0])
ax = subfig.add_subplot(1, 1, 1)

ax.plot(np.random.random(10), np.random.random(10), marker='.', linestyle='none', picker=5)

fig.canvas.mpl_connect('pick_event', on_pick)
plt.show()
```

### Actual outcome

No output.

### Expected outcome

The following should be displayed on the console when one of the markers is clicked.

\>\> pick event!

### Operating system

Windows 10

### Matplotlib Version

3.7.1

### Matplotlib Backend

Qt5Agg

### Python version

3.8.10

### Installation

pip

**Comment by tacaswell** [comment]
Thank you for the very clear report @ndeadly !  I can confirm that this is a bug.

My suspicion is that in the pick function from `Figure` (which might be inherited from `Artist`) it is not correctly offering pick events to the `SubFigure`s of the `Figure`.

Tasks:
 - trace through the code for pick events in `Figure` and verify or disprove my theory
 - if my theory is right, offer the event to all of the `SubFigure`s
 - [bonus] add a test.  We have some machinery for creating synthetic mouse events.  Either make that work or show it will me order of magnitude more code to write a test harness than the fix was.


Labeling as good first issue because it is a clear bug and there is no API design here (the pick event should fire!), but medium as it will require walking through the event handling stack and reasoning through some OO code.

**Comment by QuLogic** [comment]
Small note; this is about `Axes`, not `Axis`.

**Comment by QuLogic** [comment]
> My suspicion is that in the pick function from `Figure` (which might be inherited from `Artist`) it is not correctly offering pick events to the `SubFigure`s of the `Figure`.

I think this is correct, but maybe not for an expected reason; `Figure` inherits from `Artist`, which loops through `self.get_children()` and offers the event to children whose `.axes` attribute matches the event's Axes or any child if either is `None`. And `SubFigure`, like `Figure`, does not have a parent `Axes`, _but_ it does have an `axes` attribute which is a list of `Axes` _children_. So this never matches the event `Axes` or `None`, and `SubFigure`s never get picked.

**Comment by tsabbir96** [comment]
@tacaswell @QuLogic Hi guys, do you mind if I take a look at the issue? I would like to offer my help in solving this.

**Comment by tacaswell** [comment]
@tsabbir96 Yes, we do not assign issues, if there is not a linked PR assume the issue is up for grabs and you are free to work on it.

**Comment by turnipseason** [comment]
Hi, @tsabbir96. Are you still working on this?

**Comment by klorine28** [comment]
Hello @tacaswell I think I came up with a possible solution should I post it here or do I do a pull request? if it's the latter where in the documentation should I look in? this would be my first issue.

**Comment by rcomer** [comment]
Hi @klorine28 please go ahead and make a pull request.  Our contributors' guide is here:
https://matplotlib.org/devdocs/devel/index.html

**Comment by klorine28** [comment]
sorry to ask but in what branch should I modify the code? i know its not main but I am unsure on wich one I should do it

**Comment by rcomer** [comment]
@klorine28 see our "Workflow" guide 😀
https://matplotlib.org/devdocs/devel/development_workflow.html

**Comment by QuLogic** [comment]
This has been fixed by #27343.
